libsoftwarecontainer/unit-test/devicenode_unittest.cpp

```diff
@@ -0,0 +1,211 @@
+                                    \"mode\":  764\
+                                }";
+
+    ASSERT_TRUE(dev.parse(configJSON1));
+
+    json_t *configJSON2 = convertToJSON(config2);
+    ASSERT_TRUE(dev.parse(configJSON2));
+    ASSERT_FALSE(dev.name.empty());
+    ASSERT_EQ(dev.mode, 764);
+}
+
+/*
+ * Test that setting mode fails if name is missing
+ */
+TEST_F(DeviceNodeTest, TestConfigNoName) {
+    const std::string config = "{ \"mode\": 644 }";
+
+    json_t *configJSON = convertToJSON(config);
+
+    ASSERT_FALSE(dev.parse(configJSON));
+    ASSERT_TRUE(dev.name.empty());
+}
+
+/*
+ * Test that setting mode of bad type fails
+ */
+TEST_F(DeviceNodeTest, TestConfigBadMode) {
+    const std::string config = "{ \"name\": \"TEST_DEVICE\", \
+                                  \"mode\": \"C\"\
+                                }";
+    json_t *configJSON = convertToJSON(config);
+
+    ASSERT_FALSE(dev.parse(configJSON));
+    ASSERT_EQ(dev.mode, -1);
+}
+
+
+class MockContainer : public ContainerAbstractInterface
+{
+public :
+    MockContainer () {}
+    const char *id() const {return "";}
+    const std::string &rootFS() const {return "";}
+
+    bool initialize() {return true;}
+    bool create() {return true;}
+    bool start(pid_t *) {return true;}
+    bool stop() {return true;}
+
+    bool shutdown() {return true;}
+    bool shutdown(unsigned int) {return true;}
+
+    bool suspend() {return true;}
+    bool resume() {return true;}
+
+    bool destroy() {return true;}
+    bool destroy(unsigned int) {return true;}
+
+    bool mountDevice(const std::string &) {return true;}
+    bool createSymLink(const std::string &, const std::string &) {return true;}
+
+    bool bindMountInContainer(const std::string &, const std::string &, bool ) {return true;}
+
+    bool setEnvironmentVariable(const std::string &, const std::string &) {return true;}
+    bool setCgroupItem(std::string, std::string) {return true;}
+
+    bool execute(ExecFunction, pid_t *, const EnvironmentVariables &, int, int, int) {return false;}
+    bool execute(const std::string &,
+                 pid_t *,
+                 const EnvironmentVariables &,
+                 const std::string &,
+                 int,
+                 int,
+                 int ) {return false;}
+};
+
+
+/*
+ * Tests activate stage of the device
+ */
+TEST_F(DeviceNodeTest, Activation) {
+    std::shared_ptr<ContainerAbstractInterface> container;
+    container = std::shared_ptr<ContainerAbstractInterface> (new MockContainer());
+
+    const std::string config = "{\"name\": \"/dev/input/mouse0\", \"mode\":  622}";
+    json_t *configJSON = convertToJSON(config);
+
+    ASSERT_TRUE(dev.parse(configJSON));
+    ASSERT_EQ(dev.mode, 622);
+    ASSERT_FALSE(dev.isConfigured);
+    ASSERT_TRUE(dev.activate(container));
+    ASSERT_TRUE(dev.isConfigured);
+}
+
+
+
+struct testSetup
+{
+    Device firstConfiguration;
+    Device secondConfiguration;
```

Comment:
We can create another test to examine this. But what would have been the perks of having it ? Since this feature is provided by inherited Gateway class, it is actually on/off option. Either a gateway can be activated once or no limitation.